source/extensions/filters/http/jwt_authn/verifier.cc

```diff
@@ -0,0 +1,163 @@
+#include "extensions/filters/http/jwt_authn/verifier.h"
+
+#include "common/common/enum_to_int.h"
+#include "common/http/message_impl.h"
+#include "common/http/utility.h"
+
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtProvider;
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtRequirement;
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtRequirementAndList;
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtRequirementOrList;
+using ::Envoy::Http::LowerCaseString;
+using ::google::jwt_verify::Status;
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace JwtAuthn {
+
+void AuthVerifier::beforeVerify(const std::vector<std::string>& audiences, Http::HeaderMap& headers,
+                                Verifier::Callbacks& callback) {
+  auth_ = factory_(audiences);
```

Comment:
This is tricky part.   auth_ and callback are dynamic,  they are per request.  

I understand now why you need to create vector of matchers and verifiers per request since you hold these dynamic data in the verifier.

My suggestion is:  not to hold these dynamic data in the verifier.   Use some context to hold these dynamic data.  Verifier only hold the static data so they can be created at FilterConfig time.   

They will process context which hold the dynamic data.   context is created per-request.

Reply:
Thanks for all your feed backs. What do you think about using a visitor pattern for context and verifiers? Context would hold all the authenticator and callbacks related data, and verifier would populate the context with auths objects per requires type.

Reply:
Ideally,  the code in the Filter code should look like this:

VerifierContext ctx(headers,  this /*as Veridifer::Callback);
auto matcher = filter_config.findMatcher(ctx);
matcher->verify(ctx);

Matcher list and Verifier list only store config static data.  They can be created at FilterConfig time, not at run-time

Reply:
So, don't create the authenticator per request but, encapsulate the data in authenticator's member variables in a context object?

Reply:
I don't understand your comment about registering filter as a callback in the context. Do you want context to handle all the callbacks?

Reply:
Yes, callback is dynamic data.  It should be stored in the context, not in the verifier. That is all I want to say.

Reply:
if it is all single threaded. Wouldn't it be enough to reset all the data in authenticator per request?

Reply:
Thanks for the clarifications. Didn't see your responses when I added more questions.

Reply:
Is the question about authenticator?  I think most of its data are dynamic.  My preference is to create one for each token.   But it is up to you, if you want to move its dynamic data to ctx and re-use it for many token,  it is fine too.